File: Atcoder/DP contest/Sushi.swift
let n = Int(readLine()!)!
var arr = readLine()!.split(separator:" ").compactMap{Int($0)}
var mapNum = [1:0, 2:0, 3:0]

for a in arr {
    mapNum[a]! += 1
}

var dp:[[[Float]]] = Array(repeating:Array(repeating: Array(repeating: 0.0, count: n+1), count: n+1), count:n+1)

for i in 0...n {
    for j in 0...n {
        for k in 0...n {
            let zero = n - i - j - k
            if zero == n || zero < 0 {
                continue
            }

            var val: Float = 1.0
            if i > 0 {
                val += (i / n) * dp[i-1][j+1][k] 
            }
            if j > 0 {
                val +=  (j / n) * dp[i][j-1][k+1]
            }
            if k > 0 {
                val += (k / n) * p[i][j][k-1]
            }

            dp[i][j][k] = val / (1 - Float(zero/n))
        }
    }
}
print(String(format: "%.10f", dp[mapNum[3]!][mapNum[2]!][mapNum[1]!]))

File: Algorithms Implementations/Atcoder/DP contest/Sushi.py
n = int(input())

arr = list(map(int, input().split()))
map_num = {1: 0, 2: 0, 3: 0}

for num in arr:
    map_num[num] += 1

# dp[three][two][one] = 초밥이 1,2,3개 남은 접시가 각각 one,two,three개 남아있을 때,
# 다 먹을떄까지 필요한 operation 횟수의 기대값.
dp = [[[0 for _ in range(n + 1)] for _ in range(n + 1)] for _ in range(n + 1)]

for three in range(n + 1):
    for two in range(n + 1):
        for one in range(n + 1):
            zero = n - three - two - one

            if zero == n or zero < 0:
                continue

            val = 1
            if three > 0:
                val += (three / n) * dp[three - 1][two + 1][one]
            if two > 0:
                val += (two / n) * dp[three][two - 1][one + 1]
            if one > 0:
                val += (one / n) * dp[three][two][one - 1]

            dp[three][two][one] = val / (1 - 1.0 * zero / n)

print(dp[map_num[3]][map_num[2]][map_num[1]])
